Set ExpansionFactor to 0.06 when it's equals to 0 in the private dict of CFF fonts

calixteman opened the pull request.

calixteman committed:
Set ExpansionFactor to 0.06 when it's equals to 0 in the private dict of CFF fonts

Snuffleupagus commented on test/test_manifest.json:
```diff
@@ -7292,5 +7292,12 @@
       "link": true,
       "type": "eq"
+   },
+   {
+      "id": "issue15289",
+      "file": "pdfs/issue15289.pdf",
+      "md5": "f3d20aee398e052ac3dce7cc030f50de",
+      "rounds": 1,
+      "type": "eq"
```
It should suffice to test only one page, right?
```suggestion
   {
      "id": "issue15289",
      "file": "pdfs/issue15289.pdf",
      "md5": "f3d20aee398e052ac3dce7cc030f50de",
      "rounds": 1,
      "lastPage": 1,
      "type": "eq"
```

Snuffleupagus reviewed (approved):
That's a fascinating find, how did you figure it out?

r=me, with the comment addressed and passing tests; thank you!

calixteman commented:
> That's a fascinating find, how did you figure it out?
> 

I compared the original font I extracted from the pdf and the one we generate.
I ran `tx` on both fonts and noticed some differences in the charsets, so I manually fix the charset to have exactly the same as in the original font but nothing.
I ran `fontlint` on the generated font and noticed few errors about `BlueValues`, I fixed them but nothing. I was pretty sure that something was wrong in the top dict.
Finally I had the wonderful idea to just remove the private dict and magically it worked, then I just tried to remove the properties one by one and finally the last one (as usual) was the culprit.

calixteman commented:
/botio test

Snuffleupagus commented on test/test_manifest.json:
```diff
@@ -7292,5 +7292,13 @@
       "rounds": 1,
       "link": true,
       "type": "eq"
+   },
+   {
+      "id": "issue15289",
+      "file": "pdfs/issue15289.pdf",
+      "md5": "f3d20aee398e052ac3dce7cc030f50de",
+      "rounds": 1,
+      "lastPage": 1,
+      "type": "eq"
```
There's already a test-case for this document, hence you'll need to remove all the changes to the `test/`-folder in this patch.

calixteman commented:
/botio test

calixteman commented:
@Snuffleupagus instead of removing the property I set it to 0.06 to follow Jonathan's advice: https://github.com/mozilla/pdf.js/issues/15289#issuecomment-1374484221
Are you ok with this change ?

Snuffleupagus commented:
> Are you ok with this change ?

Yes, since I do agree that it sounds like an even safer approach here.

Given the amount of duplicates, we might want to uplift this to previous Firefox versions?

---

As a follow-up to this PR, to prevent future bugs, I do wonder if we want to add similar code when parsing "regular" Type1 fonts as well or if should we perhaps wait until an actual bug is reported?
https://github.com/mozilla/pdf.js/blob/d6984f93954a8ac9e92cc9f156f7e909d3e332de/src/core/type1_parser.js#L657

calixteman commented:
/botio makeref

calixteman commented:
/botio-windows makeref
